packages/core/src/model.ts

```diff
@@ -55,50 +62,91 @@ type EventCreators<Self> = {
     : 'An event creator must be a function';
 };
 
-type ModelCreators<Self> = {
-  events: EventCreators<Prop<Self, 'events'>>;
+type FinalEventCreators<Self> = {
+  [K in keyof Self]: Self[K] extends AnyFunction
+    ? (
+        ...args: Parameters<Self[K]>
+      ) => Compute<ReturnType<Self[K]> & { type: K }>
+    : never;
 };
 
-type FinalEventCreators<Self> = {
+type ActionCreator<
+  Self extends AnyFunction,
+  Return = ReturnType<Self>
+> = Return extends object
+  ? Return extends {
+      type: any;
+    }
+    ? "An action creator can't return an object with a type property"
+    : Self
+  : 'An action creator must return an object';
+
+type ActionCreators<Self> = {
+  [K in keyof Self]: Self[K] extends AnyFunction
+    ? ActionCreator<Self[K]>
+    : 'An action creator must be a function';
+};
+
+type FinalActionCreators<Self> = {
   [K in keyof Self]: Self[K] extends AnyFunction
     ? (
         ...args: Parameters<Self[K]>
       ) => Compute<ReturnType<Self[K]> & { type: K }>
     : never;
 };
 
+type ModelCreators<Self> = {
+  events?: EventCreators<Prop<Self, 'events'>>;
+  actions?: ActionCreators<Prop<Self, 'actions'>>;
+};
+
 type FinalModelCreators<Self> = {
   events: FinalEventCreators<Prop<Self, 'events'>>;
+  actions: FinalActionCreators<Prop<Self, 'actions'>>;
 };
 
-type EventFromEventCreators<EventCreators> = {
-  [K in keyof EventCreators]: EventCreators[K] extends AnyFunction
-    ? ReturnType<EventCreators[K]>
-    : never;
-}[keyof EventCreators];
+type EventFromEventCreators<TEventCreators> = keyof TEventCreators extends never
+  ? EventObject
+  : {
+      [K in keyof TEventCreators]: TEventCreators[K] extends AnyFunction
+        ? ReturnType<TEventCreators[K]>
+        : never;
+    }[keyof TEventCreators];
 
 export function createModel<TContext, TEvent extends EventObject>(
   initialContext: TContext
-): Model<TContext, TEvent, void>;
```

Comment:
I think there was a reason behind this `void` here - will have to take this branch for a spin locally to try few things out. I'll try my best to add some additional type tests for the desired behavior.